doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1748,31 +1757,39 @@ Controlling the proof flow
    impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
    local context.
 
-.. tacn:: contradiction
-   :name: contradiction
+.. tacn:: contradiction {? @one_term_with_bindings }
 
-   This tactic applies to any goal. The contradiction tactic attempts to
-   find in the current context (after all intros) a hypothesis that is
-   equivalent to an empty inductive type (e.g. :g:`False`), to the negation of
-   a singleton inductive type (e.g. :g:`True` or :g:`x=x`), or two contradictory
-   hypotheses :g:`P` and :g:`not P`.
+   Tries to prove the current goal by finding a contradiction.  If
+   :n:`@one_term_with_bindings` is not provided, the tactic first does a
+   :tacn:`intros`
+   then succeeds if there is a hypothesis :n:`False` in the updated context, or
+   if there is a pair of hypotheses where one is simple negation of the other
+   (i.e. :n:`P` and not :n:`~P`).
```

Comment:
Missing in this paragraph is the case where a hypothesis is the negation of "a singleton inductive type (e.g. :g:`True` or :g:`x=x`)"

Reply:
How is `x=x` a singleton inductive type?

This seems related what I ask about [below](https://github.com/coq/coq/pull/16659#discussion_r1022362456)

Reply:
"singleton inductive" means "inductive type (family) with one constructor".  `eq` has only one constructor, `eq_refl`.  So perhaps it would be more accurate to say "singleton inductive goal which can be solved simply by running `constructor`" or something like that?

Reply:
I think we should follow Jason's suggestion here.

Reply:
I think it's not helpful to refer to `constructor`.  Not as easy to understand as what I have.  (Nitpick: in a hypothesis, not in the goal.)

Reply:
@JasonGross For `contradiction`, IIUC `(1=0)` is an inductive with a single constructor (and a true, provable statement).  Can you suggest better wording that excludes `(1=0)`?  Probably it should say that `T` in `~T` must "reduce" to True (but using more precise language, e.g. exactly what reduction)?

![image](https://user-images.githubusercontent.com/1253341/209253716-fc48f48b-ddaa-477f-82b8-3c05a2af171d.png)


```
Goal ~(1=0) -> False.
intros.
contradiction.  (* No such contradiction *)
```

Reply:
There is no "reduction" that would transform `1=0` to `True` (or to anything else, really). I think having sufficiently precise language excluding this case was precisely Jason's objective when he wrote "which can be solved by running `constructor`".

Reply:
> There is no "reduction" that would transform `1=0` to `True` (or to anything else, really).

A misunderstanding.  My question asked how to exclude `(1=0)`. `~(1=0)` doesn't work for `contradiction`.  "Probably it should say that `T` in `~T` must "reduce" to True (but using more precise language, e.g. exactly what reduction)?" is about cases that work, such as `(0=0)`.

How about this?  (The "for which ..." is a bit of novel language, feel free to suggest better wording.  I wanted to indicate that `constructor` is in a different proof):

   - there is a hypothesis that is the negation of a singleton inductive type `T`
     (such as :n:`True` or :n:`x=x`) for which `Goal T. constructor.` is a complete
     proof.

Reply:
Your proposal is acceptable to me, but I think I have a preference for a slightly simpler form such as:

- there is a hypothesis that is the negation of a singleton inductive type that Coq could prove by running `constructor` (e.g. `True` or `x=x`).

IMHO it kind of makes sense to consider that Coq would create a proof that would at some point run `constructor` internally here, much like we sometimes write that Coq runs `intros` internally.

Anyway, feel free to chose your wording, mine, or something in between.

Reply:
I think I'll go with this:

   - there is a hypothesis :n:`~P` where `P` is a singleton inductive type
     (such as :n:`True` or :n:`x=x`) that `Goal P. constructor.` can prove.